File: day_7/day_7_2.rb
# frozen_string_literal: true

class NoSpaceLeftOnDevice
  def initialize(filename)
    @data = parse_device_data(filename)
    @directory_sizes = []
  end

  def run
    space_used = analyze_directory_contents
    space_needed = 30000000 - (70000000 - space_used)

    puts "The size of the smalled directory that can be deleted: #{small_directory_to_delete(space_needed)}"
  end

  def small_directory_to_delete(space_needed)
    @directory_sizes.select  {|size| size >= space_needed}.min
  end

  def analyze_directory_contents
    final_size = 0

    while line = @data.shift
      line_parts = line.split
      case line_parts
        in ['$', 'cd', '..']
          @directory_sizes << final_size
          return final_size
        in ['$', 'cd', _]
          final_size += analyze_directory_contents
        in [file_size, _]
          final_size += file_size.to_i
      end
    end

    final_size
  end

  def parse_device_data(filename)
    File.read(filename).lines
  end

end

NoSpaceLeftOnDevice.new(ARGV[0] || 'input.txt').run

File: day_7/day_7_2.rs
use std::collections::VecDeque;
use std::env;

struct NoSpaceLeftOnDevice {
    data: VecDeque<String>,
    directory_sizes: Vec<i32>,
}

impl NoSpaceLeftOnDevice {
    fn new(filename: &str) -> NoSpaceLeftOnDevice {
        let data = Self::parse_device_data(filename).unwrap();
        NoSpaceLeftOnDevice {
            data: data,
            directory_sizes: Vec::<_>::new(),
        }
    }

    fn parse_device_data(filename: &str) -> Option<VecDeque<String>> {
        let data = std::fs::read_to_string(filename).ok()?;
        Some(data.lines().map(String::from).collect())
    }

    fn run(&mut self) {
        let space_used = self.analyze_directory_contents();
        let space_needed = 30000000 - (70000000 - space_used);

        println!(
            "The size of the smalled directory that can be deleted: #{}",
            self.small_directory_to_delete(space_needed)
        );
    }

    fn small_directory_to_delete(&self, space_needed: i32) -> i32 {
        self.directory_sizes
            .iter()
            .filter(|&size| size >= &space_needed)
            .map(|size| *size)
            .min()
            .unwrap()
    }

    fn analyze_directory_contents(&mut self) -> i32 {
        let mut final_size = 0;

        while let Some(line) = self.data.pop_front() {
            let line_parts = line.split(" ").collect::<Vec<&str>>();
            match line_parts.as_slice() {
                ["$", "cd", ".."] => {
                    self.directory_sizes.push(final_size);
                    return final_size;
                }
                ["$", "cd", _] => {
                    final_size += self.analyze_directory_contents();
                }
                ["$", "ls"] => {}
                ["dir", _] => {}
                [file_size, _] => {
                    final_size += file_size.parse::<i32>().unwrap();
                }
                _ => {}
            }
        }
        final_size
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut filename = "input.txt";
    if let Some(arg) = args.get(1) {
        filename = arg;
    }

    NoSpaceLeftOnDevice::new(filename).run();
}
